### PR TITLE
LooseObjectsStep (Part 2)

### PR DESCRIPTION
**Pull request by jeschu1**
- Pack loose objects into a Packfile

Update to let Stolee know how things are looking.  Don't worry about a review yet everyone else.

Next steps for me are to add the unit tests and make sure I'm happy with all the logging.

### CODE REVIEW

File: GVFS/GVFS.Common/Git/GitProcess.cs [hunk 1]
```diff
@@ -373,6 +373,16 @@ public Result UnpackObjects(Stream packFileStream)
                 null);
         }
 
+        public Result PackObjects(string filenamePrefix, string gitObjectsDirectory, Action<StreamWriter> packFileStream)
+        {
+            string packFilePath = Path.Combine(gitObjectsDirectory, "pack", filenamePrefix);
+            return this.InvokeGitAgainstDotGitFolder(
+                $"pack-objects {packFilePath} --non-empty",
```
Comment (derrickstolee) [inline GVFS/GVFS.Common/Git/GitProcess.cs]: We may want a `-q` here to avoid progress output. But it depends on what the logs look like.

File: GVFS/GVFS.Common/Maintenance/LooseObjectsStep.cs [hunk 2]
```diff
@@ -12,7 +12,8 @@ namespace GVFS.Common.Maintenance
     public class LooseObjectsStep : GitMaintenanceStep
     {
         public const string LooseObjectsLastRunFileName = "loose-objects.time";
-        private readonly bool forceRun;
+        public int LooseObjectsPutIntoPackFile = 0;
```
Comment (derrickstolee) [inline GVFS/GVFS.Common/Maintenance/LooseObjectsStep.cs]: `{ get; private set; }`. Initializes to zero anyway, but want to avoid public setters.

File: GVFS/GVFS.Common/Maintenance/LooseObjectsStep.cs [hunk 3]
```diff
@@ -22,26 +23,63 @@ public LooseObjectsStep(GVFSContext context, bool requireCacheLock, bool forceRu
 
         public override string Area => nameof(LooseObjectsStep);
 
+        public virtual int MaxLooseObjectsInPack => 1000;
```
Comment (derrickstolee) [inline GVFS/GVFS.Common/Maintenance/LooseObjectsStep.cs]: Some math to do here. How long does it take to pack 1000 objects? 10,000? 100,000? If we are running this only once a week, then we probably want this to be _at least_ 100,000 due to over-hydrating causes 2.1 million blob downloads. Unless there is a reason why that would be too many objects.

File: GVFS/GVFS.FunctionalTests/Tests/EnlistmentPerFixture/LooseObjectStepTests.cs [hunk 4]
```diff
@@ -73,6 +98,12 @@ private void DeleteFiles(List<string> filePaths)
             }
         }
 
+        private int CountPackFiles()
+        {
+            string[] files = Directory.GetFiles(this.PackRoot);
+            return files.Where(file => string.Equals(Path.GetExtension(file), ".pack", System.StringComparison.OrdinalIgnoreCase)).Count();
```
Comment (derrickstolee) [inline GVFS/GVFS.FunctionalTests/Tests/EnlistmentPerFixture/LooseObjectStepTests.cs]: split LINQ lines across multiple lines. No need to have the `files` variable, as we can chain LINQ:

```
return Directory.GetFiles(this.PackRoot)
                 .Where(file => string.Equals(Path.GetExtension(file), ".pack", System.StringComparison.OrdinalIgnoreCase))
                 .Count();
```

Comment (derrickstolee) [reply GVFS/GVFS.FunctionalTests/Tests/EnlistmentPerFixture/LooseObjectStepTests.cs]: and use `this.Context.FileSystem` instead of `Directory` directly?

Comment (jeschu1) [reply GVFS/GVFS.FunctionalTests/Tests/EnlistmentPerFixture/LooseObjectStepTests.cs]: Cool on the linq, since we're in a functional test I don't believe we have this.Context

File: GVFS/GVFS.Common/Git/GitProcess.cs [hunk 5]
```diff
@@ -596,6 +606,7 @@ public Process GetGitProcess(string command, string workingDirectory, string dot
                         }
 
                         writeStdIn?.Invoke(this.executingProcess.StandardInput);
+                        this.executingProcess.StandardInput.Close();
```
Comment (jrbriggs) [inline GVFS/GVFS.Common/Git/GitProcess.cs]: Why is this change necessary?

File: GVFS/GVFS.Common/Maintenance/LooseObjectsStep.cs (hunk 3 above)
Comment (jrbriggs) [reply GVFS/GVFS.Common/Maintenance/LooseObjectsStep.cs]: Agreed. Also consider the size of a typical hourly prefetch pack or a daily prefetch pack. How many objects do we see there?

File: GVFS/GVFS.Common/Maintenance/LooseObjectsStep.cs [hunk 6]
```diff
@@ -22,26 +23,63 @@ public LooseObjectsStep(GVFSContext context, bool requireCacheLock, bool forceRu
...
 
             return count;
         }
 
+        public void PackLooseObjects(StreamWriter streamWriter)
+        {
+            // Find looseObjects
+            foreach (string directoryPath in this.Context.FileSystem.EnumerateDirectories(this.Context.Enlistment.GitObjectsRoot))
+            {
+                string directoryName = directoryPath.TrimEnd(Path.DirectorySeparatorChar).Split(Path.DirectorySeparatorChar).Last();
```
Comment (jrbriggs) [inline GVFS/GVFS.Common/Maintenance/LooseObjectsStep.cs]: `System.IO.Directory.GetParent()`?

File: GVFS/GVFS.Common/Maintenance/LooseObjectsStep.cs [hunk 7]
```diff
@@ -22,26 +23,63 @@ public LooseObjectsStep(GVFSContext context, bool requireCacheLock, bool forceRu
...
+
+                if (GitObjects.IsLooseObjectsDirectory(directoryName))
+                {
+                    string[] looseObjectFileNamesInDir = this.Context.FileSystem.GetFiles(Path.Combine(this.Context.Enlistment.GitObjectsRoot, directoryPath), "*");
+
+                    foreach (string file in looseObjectFileNamesInDir)
+                    {
+                        if (this.LooseObjectsPutIntoPackFile == this.MaxLooseObjectsInPack)
```
Comment (jrbriggs) [inline GVFS/GVFS.Common/Maintenance/LooseObjectsStep.cs]: Maybe test >= rather than exact equality, in case we decide one day we'll add in batches.

File: GVFS/GVFS.Common/Maintenance/LooseObjectsStep.cs [hunk 8]
```diff
@@ -22,26 +23,63 @@ public LooseObjectsStep(GVFSContext context, bool requireCacheLock, bool forceRu
...
+                        }
+
+                        this.LooseObjectsPutIntoPackFile++;
+                        streamWriter.Write(this.GetLooseObjectGuid(directoryName, file) + "\n");
+                    }
+                }
+            }
+        }
+
+        public string GetLooseObjectGuid(string directoryName, string filePath)
```
Comment (jrbriggs) [inline GVFS/GVFS.Common/Maintenance/LooseObjectsStep.cs]: This is an oid, not a guid, right?

File: GVFS/GVFS.Common/Maintenance/LooseObjectsStep.cs [hunk 9]
```diff
@@ -22,26 +23,63 @@ public LooseObjectsStep(GVFSContext context, bool requireCacheLock, bool forceRu
...
+        {
+            return directoryName + Path.GetFileName(filePath);
+        }
+
+        public void CreateLooseObjectsPackFile()
+        {
+            GitProcess.Result result = this.RunGitCommand((process) => process.PackObjects("from-loose", this.Context.Enlistment.GitObjectsRoot, this.PackLooseObjects));
```
Comment (jrbriggs) [inline GVFS/GVFS.Common/Maintenance/LooseObjectsStep.cs]: This could use some newlines.

File: GVFS/GVFS.Common/Maintenance/LooseObjectsStep.cs [hunk 10]
```diff
@@ -65,15 +103,18 @@ protected override void PerformMaintenance()
...
 
                     EventMetadata metadata = new EventMetadata();
                     metadata.Add("GitObjectsRoot", this.Context.Enlistment.GitObjectsRoot);
-                    metadata.Add("StartingCount", beforeCount);
-                    metadata.Add("EndingCount", afterCount);
-                    metadata.Add("RemovedCount", beforeCount - afterCount);
+                    metadata.Add("StartingCount", beforeLooseObjectsCount);
+                    metadata.Add("EndingCount", afterLooseObjectsCount);
+                    metadata.Add("RemovedCount", beforeLooseObjectsCount - afterLooseObjectsCount);
+                    metadata.Add(nameof(this.LooseObjectsPutIntoPackFile), this.LooseObjectsPutIntoPackFile);
```
Comment (jrbriggs) [inline GVFS/GVFS.Common/Maintenance/LooseObjectsStep.cs]: nit: let's not make this nameof here, so that refactoring doesn't change what we look for in the data exhaust streams.

File: GVFS/GVFS.Common/Maintenance/LooseObjectsStep.cs (hunk 8 above)
Comment (jeschu1) [reply GVFS/GVFS.Common/Maintenance/LooseObjectsStep.cs]: Updating to just Id as per @derrickstolee  that is most correct

File: GVFS/GVFS.Common/Git/GitProcess.cs (hunk 5 above)
Comment (derrickstolee) [reply GVFS/GVFS.Common/Git/GitProcess.cs]: `git pack-objects` listens to stdin until receiving an EOF. This is different from the other commands we've used in the past, which cared about an empty line.

File: GVFS/GVFS.Common/Maintenance/LooseObjectsStep.cs (hunk 3 above)
Comment (derrickstolee) [reply GVFS/GVFS.Common/Maintenance/LooseObjectsStep.cs]: Keep in mind that the daily/hourly prefetch packs contain commits and trees, and we expect this step to create packs from blobs. The blobs will be much bigger.

File: GVFS/GVFS.Common/Maintenance/LooseObjectsStep.cs (hunk 6 above)
Comment (derrickstolee) [reply GVFS/GVFS.Common/Maintenance/LooseObjectsStep.cs]: or rather, `Path.GetDirectory()`?

File: GVFS/GVFS.Common/Maintenance/LooseObjectsStep.cs (hunk 8 above)
Comment (derrickstolee) [reply GVFS/GVFS.Common/Maintenance/LooseObjectsStep.cs]: OID.

Comment (derrickstolee) [reply GVFS/GVFS.Common/Maintenance/LooseObjectsStep.cs]: right, ObjectId

File: GVFS/GVFS.Common/Git/GitProcess.cs (hunk 5 above)
Comment (jrbriggs) [reply GVFS/GVFS.Common/Git/GitProcess.cs]: Got it. It's a harmless change for other commands but necessary for `git pack-objects`. Should we have a comment on why the line is needed?

File: GVFS/GVFS.Common/Maintenance/LooseObjectsStep.cs (hunk 3 above)
Comment (jrbriggs) [reply GVFS/GVFS.Common/Maintenance/LooseObjectsStep.cs]: Fair point. Either way I think 1000 is too small by an order of magnitude or two.

File: GVFS/GVFS.Common/Maintenance/LooseObjectsStep.cs (hunk 6 above)
Comment (jeschu1) [reply GVFS/GVFS.Common/Maintenance/LooseObjectsStep.cs]: Are you suggesting instead of Split(Path.DirectorySeparatorChar).Last(), substring using GetParent()?

Comment (jeschu1) [reply GVFS/GVFS.Common/Maintenance/LooseObjectsStep.cs]: If we do a substring @derrickstolee had the suggestion of 
directoryPath.TrimEnd(Path.DirectorySeparatorChar).Substring(directoryPath.LastIndexOf(separator) + 1)

Shame there isn't a builtin function to do this but I'm not really bothered with which permuation we go with.

Comment (jeschu1) [reply GVFS/GVFS.Common/Maintenance/LooseObjectsStep.cs]: @derrickstolee I think I can use that to simplify

**Comment by jeschu1** [comment]
Sorry guys it looks like i made the original branch here instead of my fork.  I'll delete the branch in the MIcrosoft project and am opening a new PR.

File: GVFS/GVFS.Common/Maintenance/LooseObjectsStep.cs (hunk 3 above)
Comment (jeschu1) [reply GVFS/GVFS.Common/Maintenance/LooseObjectsStep.cs]: Went with 100,000 for now
